### PR TITLE
feat: [CDS-83902]: send env gitmetadata to use it for infrastructure apis in pipeline studio

### PR DESCRIPTION
**Pull request by sonaligoyal-harness**
### Summary

This change is supported only for pipeline studio.
For infrastructure yaml metadata, we need reference of environment repo and branch . So for each envRef branch and repo is stored in local state and passed further to use it for infrastructures.
 
 - Tested single env single infra
 - Tested multi env multi infra
 - Test remote env and inline infra 
 - Tested remote pipeline, remote env , remote infra

#### Screenshots

single env single infra

https://github.com/harness/harness-core-ui/assets/98508399/c71a3a12-f8be-46da-85ae-7cdda8513093

multi env multi infra

https://github.com/harness/harness-core-ui/assets/98508399/6ac21bb3-e20c-4ead-ac0a-85bf5747fb77

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
- MessageMetadata: `trigger messagecheck`
- GitLeaks: `trigger gitLeaks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

### CODE REVIEW

**Comment by sonaligoyal-harness** [comment]
trigger messagecheck

File: src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployInfrastructure/useGetInfrastructuresData.ts [hunk 1]
```diff
@@ -125,6 +124,9 @@ export function useGetInfrastructuresData({
     lazy: sortedInfrastructureIdentifiers.length === 0 || isGitXEnabledForInfras
   })
 
+  const environmentsGitMetadata = JSON.parse(localStorage.getItem('environmentGitDetails') as string) || {}
```
Comment (chetanvekariya-harness) [inline src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployInfrastructure/useGetInfrastructuresData.ts:127]: Can't we use useLocalStorage hook here as well?

File: src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployEnvironment/useGetEnvironmentsData.ts [hunk 2]
```diff
@@ -150,6 +150,11 @@ export function useGetEnvironmentsData({
   })
 
   const loading = loadingEnvironmentsList || loadingEnvironmentsData || loadingEnvironmentsDataV2
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [_envGitDetails, setEnvGitDetails] = useLocalStorage<Record<string, EnvironmentGitMetaDataForInfrastructure>>(
```
Comment (rameshx) [inline src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployEnvironment/useGetEnvironmentsData.ts:154]:
```suggestion
  const [, setEnvGitDetails] = useLocalStorage<Record<string, EnvironmentGitMetaDataForInfrastructure>>(
```

File: src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployEnvironment/useGetEnvironmentsData.ts [hunk 3]
```diff
@@ -245,6 +250,27 @@ export function useGetEnvironmentsData({
         })
       }
 
+      if (!isEmpty(_environmentsData)) {
+        const mappedEnvGitMetaData = sortedEnvIdentifiers.reduce(
+          (result: Record<string, EnvironmentGitMetaDataForInfrastructure>, envRef) => {
+            // Find the corresponding data for the current envRef
+            const envData = _environmentsData?.find(item => getScopedValueFromDTO(item.environment) === envRef)
+            // If data is found, add it to the result array
```
Comment (rameshx) [inline src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployEnvironment/useGetEnvironmentsData.ts:258]: Nit: `result` is not an array

File: src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployInfrastructure/useGetInfrastructuresData.ts (hunk 1 above)
Comment (rameshx) [reply src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployInfrastructure/useGetInfrastructuresData.ts:127]: +1

**Review by rameshx — APPROVED** [review]

**Review by chetanvekariya-harness — APPROVED** [review]

**Review by vikrant-harness — APPROVED** [review]

**Pull request closed without merging**